### ISSUE TITLE
Add high level autograd functions

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#30632]
It would be very useful to have high level autograd functions such as:
- jacobian
- hessian
- jvp
- vjp

The questions are:
- Where should they go? `torch.autograd.??`
- What kind of api we want to provide: should they be function to function mapping? Or should they be function + value to value mappings?

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by zou3519** [comment]
> * Where should they go? `torch.autograd.??`

Seems reasonable. What were your concerns for putting them in autograd or putting them not in autograd? Right now our autograd is reverse-mode but I think it makes sense for autograd to be able to do different things.

> * What kind of api we want to provide: should they be function to function mapping? Or should they be function + value to value mappings?

I kind of like jax-style function-to-function mappings because they make the code feel more composable. Realistically speaking one potential way to determine which one to do would be to write some code in function-to-function style, and rewrite it again in (function + value to value) style, and then see which one looks better. I've been told that [neural tangents](https://github.com/google/neural-tangents) is a good example because it involves composition of a vjp with a jvp function somewhere.

**Comment by albanD** [comment]
- My concern was more how to name this new "submodule" of autograd. But that is not a urgent issue.

- I do like the function-to-function style as well, but we support many more operations, including arbitrary side effects. So our function-to-function version would be much weaker than theirs. But maybe we're happy with this? I agree that writing both these things with each style should give good signal on the api.

**Comment by zou3519** [comment]
> * I do like the function-to-function style as well, but we support many more operations, including arbitrary side effects. So our function-to-function version would be much weaker than theirs.

That's a great point. I don't know how in-place ops and views would affect this -- how would our function-to-function version turn out weaker?

Since this is a new API, one thing we could do (maybe extreme) is to ban side effects. If we were redesigning autograd/pytorch from scratch, would we have arbitrary side effects?

**Comment by ezyang** [comment]
What's wrong with function-to-function mappings with in-place operations? I don't think there is a problem here.

**Comment by albanD** [comment]
My concerns are:
- Is it the most pythonic way to do this in pytorch? It will be a brand new paradigm compared to the current autograd functions.
- in-place are fine but I was more thinking about functions with state. You can end up with a function `f` where multiple calls to `f(x)` in a row will give different results. But I guess this is ok.
- the function-to-function is nice to easily compose many of them. But in our case, the composition might not be as nice as the ones done in jax for example. This could give surprising performance to the user?

**Comment by ezyang** [comment]
> Is it the most pythonic way to do this in pytorch? It will be a brand new paradigm compared to the current autograd functions.

I mean, there is always a possibility that we should support both modes. At the end of the day, the primitive operation we provide is the imperative (not function-to-function) version, so the imperative version of the API will be more "transparent" in some sense. Another way to put it is, there are two orthogonal issues here: first is adding function-function operations for our existing backward operations, and the second is adding things like hessian to the library.

> in-place are fine but I was more thinking about functions with state. You can end up with a function f where multiple calls to f(x) in a row will give different results. But I guess this is ok.

You're going to have to say more. For example, suppose I run `primals, f_vjp = jax.vjp(f, x, y)` where x has side effects. The side effects will get executed once during the vjp call; the `f_vjp` backwards, however, is a pure function and won't have any side effects. If I run f again with different primals, of course the side effects will happen again differently!

> the function-to-function is nice to easily compose many of them. But in our case, the composition might not be as nice as the ones done in jax for example. This could give surprising performance to the user?

There is a performance consideration here, though not the one you're thinking about. In a JAX style API where we return a `f_vjp` function, we cannot assume that `f_vjp` is a one-shot function (e.g., that it will be called only once). This means that we're not permitted to clear `grad_fn` when it's called, aka we have to call backward with `retain_graph=True`. Furthermore, we don't know if the user is going to compose this with other differentiation, so we also have to `create_graph=True`. So this could end up being slower if the user is not careful. But this is a sort of intrinsic problem to eager mode, and I don't think there is actually a way around it here.

**Comment by albanD** [comment]
I would agree with you that given the current state, an imperative version will be necessary and a function-to-function version will just be using function closures on top of the imperative version.

I'll send a proposal for this.

**Comment by albanD** [comment]
Here is the api I propose for these functions placed under the `torch.autograd.higher` module:

```python

def vjp(func, inputs, v=None, create_graph=False):
    r"""Function that computes the dot product between a vector ``v`` and the Jacobian of
    the given function at the point given by the inputs.

    Args:
        func (function): a Python function that takes Tensor inputs and returns
            a sequence of Tensor or a Tensor.
        inputs (sequence of Tensor or Tensor): inputs to the function.
        v (sequence of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
            same size as the output of ``func``. This argument is optional when
            ``func``'s output contains a single element and it will be set as a Tensor
            containing a single ``1``.
        create_graph (bool, optional): If ``True``, both the output and result will be
            computed in a differentiable way.
            Defaults to ``False``.

    Returns:
        output (sequence of Tensor): output of ``func(inputs)``
        result (sequence of Tensor): result of the dot product with the same shape
            as the inputs.

    Example::

        >>> def reducer(x):
        ...   return x.sum(dim=1)
        >>> inputs = torch.rand(4, 4)
        >>> v = torch.ones(4)
        >>> vjp(reducer, inputs, v)
        (tensor([1.7465, 1.9952, 2.8585, 1.7456]),
         tensor([[1., 1., 1., 1.],
                 [1., 1., 1., 1.],
                 [1., 1., 1., 1.],
                 [1., 1., 1., 1.]])

        >>> vjp(reducer, inputs, v, create_graph=True)
        (tensor([1.7465, 1.9952, 2.8585, 1.7456], grad_fn=<SumBackward1>),
         tensor([[1., 1., 1., 1.],
                 [1., 1., 1., 1.],
                 [1., 1., 1., 1.],
                 [1., 1., 1., 1.]], grad_fn=<ExpandBackward>)

        >>> def adder(x, y):
        ...   return 2 * x + 3 * y
        >>> inputs = (torch.rand(2), torch.rand(2))
        >>> v = torch.ones(2)
        >>> vjp(adder, inputs, v)
        (tensor([3.6616, 4.6449]),
         (tensor([2., 2.]), tensor([3., 3.])))
    """

    raise NotImplementedError("TODO")


def jvp(func, inputs, v=None, create_graph=False):
    r"""Function that computes the dot product between  the Jacobian of
    the given function at the point given by the inputs and a vector ``v``.

    Args:
        func (function): a Python function that takes Tensor inputs and returns
            a sequence of Tensor or a Tensor.
        inputs (sequence of Tensor): inputs to the function.
        v (sequence of Tensor): The vector that will multiply the Jacobian. Must be the
            same size as the input of ``func``. This argument is optional when
            ``func``'s input contains a single element and it will be set as a Tensor
            containing a single ``1``.
        create_graph (bool, optional): If ``True``, both the output and result will be
            computed in a differentiable way.
            Defaults to ``False``.

    Returns:
        output (sequence of Tensor): output of ``func(inputs)``
        result (sequence of Tensor): result of the dot product with the same shape
            as the output.

    Example::

        >>> def reducer(x):
        ...   return x.sum(dim=1)
        >>> inputs = torch.rand(4, 4)
        >>> v = torch.ones(4, 4)
        >>> vjp(reducer, inputs, v)
        (tensor([1.7465, 1.9952, 2.8585, 1.7456]),
         tensor([4., 4., 4., 4.])

        >>> vjp(reducer, inputs, v, create_graph=True)
        (tensor([1.7465, 1.9952, 2.8585, 1.7456], grad_fn=<SumBackward1>),
         tensor([4., 4., 4., 4.], grad_fn=<SumBackward1>)

        >>> def adder(x, y):
        ...   return 2 * x + 3 * y
        >>> inputs = (torch.rand(2), torch.rand(2))
        >>> v = (torch.ones(2), torch.ones(2))
        >>> vjp(adder, inputs, v)
        (tensor([3.6616, 4.6449]),
         tensor([5., 5.]))
    """

    raise NotImplementedError("TODO")


def jacobian(func, inputs, create_graph=False):
    r"""Function that computes the Jacobian of a given function.

    Args:
        func (function): a Python function that takes Tensor inputs and returns
            a sequence of Tensor or a Tensor.
        inputs (sequence of Tensor): inputs to the function.
        create_graph (bool, optional): If ``True``, the Jacobian will be computed in
            a differentiable manner.
            Defaults to ``False``.

    Returns:
        Jacobian (Tensor or sequence of sequence of Tensor) if there are a single input
            and output, this will be a single Tensor containing the Jacobian for the
            linearized inputs and output. If any of the two is a tuple, then the Jacobian
            will be a list of tuples where ``Jacobian[i][j]`` will contain the Jacobian
            of the linearized ``i``th output and ``j``th input.

    Example::

        >>> def exp_reducer(x):
        ...   return x.exp().sum(dim=1)
        >>> inputs = torch.rand(4, 4)
        >>> jacobian(exp_reducer, inputs)
        tensor([[2.5312, 1.9348, 2.1872, 1.5616, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
                 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
                [0.0000, 0.0000, 0.0000, 0.0000, 1.6283, 1.2821, 1.2146, 1.6624, 0.0000,
                 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
                [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 2.3369,
                 1.1236, 1.0116, 1.1355, 0.0000, 0.0000, 0.0000, 0.0000],
                [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
                 0.0000, 0.0000, 0.0000, 1.2675, 1.5187, 2.6350, 1.6487]])

        >>> jacobian(exp_reducer, inputs, create_graph=True)
        tensor([[2.5312, 1.9348, 2.1872, 1.5616, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
                 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
                [0.0000, 0.0000, 0.0000, 0.0000, 1.6283, 1.2821, 1.2146, 1.6624, 0.0000,
                 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
                [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 2.3369,
                 1.1236, 1.0116, 1.1355, 0.0000, 0.0000, 0.0000, 0.0000],
                [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
                 0.0000, 0.0000, 0.0000, 1.2675, 1.5187, 2.6350, 1.6487]],
               grad_fn=<ViewBackward>)

        >>> def exp_adder(x, y):
        ...   return 2 * x.exp() + 3 * y
        >>> inputs = (torch.rand(2), torch.rand(2))
        >>> jacobian(exp_adder, inputs)
        [(tensor([[2.1747, 0.0000],
                  [0.0000, 2.3715]]),),
         (tensor([[3., 0.],
                  [0., 3.]]),)]
    """

    raise NotImplementedError("TODO")


def hessian(func, inputs, create_graph=False):
    r"""Function that computes the Hessian of a given scalar function.

    Args:
        func (function): a Python function that takes Tensor inputs and returns
            a Tensor with a single element.
        inputs (sequence of Tensor): inputs to the function.
        create_graph (bool, optional): If ``True``, the Hessian will be computed in
            a differentiable manner.
            Defaults to ``False``.

    Returns:
        Hessian (Tensor or sequence of sequence of Tensor) if there are a single input,
            this will be a single Tensor containing the Hessian for the linearized input.
            If it is a tuple, then the Hessian will be a list of tuples where
            ``Hessian[i][j]`` will contain the Hessian of the linearized ``i``th input
            and ``j``th input.

    Example::

        >>> def pow_reducer(x):
        ...   return x.pow(3).sum()
        >>> inputs = torch.rand(2, 2)
        >>> hessian(pow_reducer, inputs)
        tensor([[1.4557, 0.0000, 0.0000, 0.0000],
                [0.0000, 5.3762, 0.0000, 0.0000],
                [0.0000, 0.0000, 5.8773, 0.0000],
                [0.0000, 0.0000, 0.0000, 4.0989]])

        >>> hessian(pow_reducer, inputs, create_graph=True)
        tensor([[1.4557, 0.0000, 0.0000, 0.0000],
                [0.0000, 5.3762, 0.0000, 0.0000],
                [0.0000, 0.0000, 5.8773, 0.0000],
                [0.0000, 0.0000, 0.0000, 4.0989]], grad_fn=<ViewBackward>)


        >>> def pow_adder_reducer(x, y):
        ...   return (2 * x.pow(2) + 3 * y.pow(2)).sum()
        >>> inputs = (torch.rand(2), torch.rand(2))
        >>> hessian(pow_adder_reducer, inputs)
        [(tensor([[1.0049, 0.0000],
                  [0.0000, 2.0446]]),
          tensor([[0., 0.],
                  [0., 0.]])),
         (tensor([[0., 0.],
                  [0., 0.]]),
          tensor([[ 4.5231,  0.0000],
                  [ 0.0000, 13.3150]]))]
    """

    raise NotImplementedError("TODO")
```

cc @zdevito @apaszke

**Comment by ezyang** [comment]
@albanD If I pass in inputs that do not require grad what are the intended semantics ;)

**Comment by albanD** [comment]
My idea here is that contrary to the existing api. You just ask for an output value. You don't need to worry about how the autograd get these values. So the idea:
- If the input does not require gradients, we will change that internally without changing the value given by the user.
- If the input already does require gradients, then we just use it. When we `create_graph=True`, this graph will still be attached to the given inputs in this case.

Note that asking for `create_graph=True` when none of the inputs or `v` requires gradients will raise a warning.

**Comment by zakizhou** [comment]
Any progress?

**Comment by albanD** [comment]
I am waiting for comments on the proposed API above and will start implementing it next month. Feel free to add any feedback you have.

**Comment by yaroslavvb** [comment]
1. I'm wondering if it makes sense to allow arbitrary Hessians (not just scalar valued)? It's not much harder to implement -- can be a for loop over "single-output Hessian" and would be in line with wider convention (ie Jax does not restrict itself to scalar output Hessians). Rank of the Hessian would be the rank of output dimension + rank of input dimension + rank of input dimension. So for scalar valued function that accepts a vector, Hessian would be a matrix.

2. Hessian is symmetric so multiplying on the left or on the right gives the same result. Unlike jvp and vjp, I've only seen people refer to hvp, perhaps vhp is not needed? Non-scalar output would correspond to a batched Hessian vector product: `einsum('oij->oi', hess, vec)`

3. A gotcha of current indexing scheme is that it makes a distinction between lists and tensors:

Imagine refactoring your code and input/output spec to use `[tensor(x), tensor(y)]` instead of `tensor([x, y])`, you would then have to change your Jacobian indexing from `[i][j][k][l]` to `[i][k][j][l]`.

In other words, introducing lists/tuples forces you to interleave your spec -- your sequence of indices now have following roles "position in output->position in input->position in output->position in input". Meanwhile, keeping things as tensors, index roles are "output, output, input, input"

Some notes on how to keep sanity with derivative-indexing -- https://docs.google.com/document/d/1Sg5XG9Nh3s5N50m3PTv_qexrqnTVddmkTvYiyj24uG4/edit#heading=h.pa8ey14bqxps

**Comment by EricSpeidel** [comment]
Have you seen this [ICLR 2020 paper](https://openreview.net/forum?id=BJlrF24twB)?
The authors provide a [PyTorch implementation](https://github.com/f-dangel/backpack), too.

**Comment by albanD** [comment]
Hi,

Yes I saw this package. It is indeed very interesting but quite orthogonal to what is proposed here. They use the engine to compute quantities that the autograd engine cannot compute natively (using hooks mainly). This proposal would be using the vanilla autograd engine for all computations (only use gradients and gradients of gradients).

**Comment by albanD** [comment]
Update on the above API.
To echo the discussion that happened in this issue: https://github.com/pytorch/pytorch/issues/32576
The current autograd api does program differentiation. This means that everything that happens from the creation of the input to the backward will influence the backward.
This new API will be different in the sense that a function is specified and only that function will be differentiated. This means that we expect the inputs to be leafs.
This will force the user to properly specify inputs and will allow a simpler implementation (we don't try to allow the user to implicitly capture Tensors that require gradients as we do in checkpoint for example).

**Comment by zou3519** [comment]
Is the plan to call detach() on the inputs the user provides automatically or to enforce that every tensor being used is a leaf tensor? I'm curious if we want to limit inputs to being flat lists of tensors (that makes it easier to call detach() or check that the objects are actually leaf tensors) or let them be arbitrary (at which point it might be difficult to find tensors).

**Comment by ezyang** [comment]
Well, regardless of the detaching, you still need to identify all the input tensors, right?

**Comment by albanD** [comment]
Yes, we need to identify all the input Tensors.
The idea is that the given function `f` take **only** (floating type) Tensors as input (not list). We cannot compute gradients wrt other objects anyway. You can use `*args` if you want a flexible number of inputs. And if your function takes other non-Tensor arguments, you will need to use a lambda/wrapper function.

All the inputs will have to be leafs. And yes, we will use detach to avoid any involuntary sharing of objects that can lead to https://github.com/pytorch/pytorch/issues/32576 for example.

**Comment by albanD** [comment]
@yaroslavvb thanks for the detailed note about indexing.
The problem with the consistent indexing you propose is for functions like: `c = f(a, b)` where each Tensor a, b and c are 1D of size respectively A, B and C (note that A can be different than B).
- If you linearize both inputs into a single Tensor of size `A + B`, then it works fine, but I would argue that this is not very clear for the user to index at `i + A` to get the `i`th entry for b.
- You can consider your input as 2D where the first dimension index the argument number and the second index the position in that argument. In that case, because A and B can be different, we are forced to have the first dimension left-most and stored in a list/tuple.

Note that Tensorflow and jax does the same thing of returning a list or list of list that match the argument indexing. And each element in there correspond to the Jacobian for one output and one input:
```python
# Jax example (it only allows a single input to be differentiated)
import jax
import numpy as np

def f(x):
  return x, jax.numpy.asarray([x[0], 5*x[2], 4*x[1]**2 - 2*x[2], x[2] * jax.numpy.sin(x[0])])

res = jax.jacfwd(f)(np.array([[1., 2., 3.], [1., 1., 1.]]))
print(list(a.shape for a in res))
# [(2, 3, 2, 3), (4, 3, 2, 3)]
```

Does that make sense?

**Comment by yaroslavvb** [comment]
Yes, makes sense. Flattening is bad. A!=B forces that dimension to be a tuple/list instead of Tensor, which I think is OK, since Jax does this already. Users can work around it in performance-intensive cases now, and in future it could be optimized with some kind of ragged tensor support.

Thinking more about it, the proposed API is quite close to "left-to-right" indexing already, with main difference is due to this part:
` If any of the two is a tuple, then the Jacobian will be a list of tuples.`

This inserts an unnecessary singleton dimension in the case when only one is a tuple. I think it's better to not have this extra dimension.

At a high level, the point is to treat different implementations of the same structure as close as possible. Tensors x0, x1, x2 below have the same structure and the user may not remember difference between them.
```
x0 = [[1.,2.],[3.,4.]]
x1 = (torch.tensor([1, 2]), torch.tensor([3, 4]))
x2 = torch.stack(x1)
```

Below is an example of how Jax treats multiple arguments, which I think it's good to follow since it's in the spirit of easy to remember indexing. IE, calling Jacobian twice on `f1(x1, x2)` and `f2([x1, x2])` gives result with same depth and structure -- `result1[i][j][k][l]=result2[i][j][k][l]` (colab - [jax derivative indexing](https://colab.research.google.com/drive/1NamHYhlvjN7baslRkXz6kXRAAM69Ek9y#scrollTo=MYtSkIt-yl3t))

```
# least-squares function, f'(x) = x
def f(x):
  result = 0
  for row in x:
    for val in row:
      result += val*val*val
  return result/6

# same as above, but each row is a separate argument
def f_split(row1, row2):
  result = 0
  for row in (row1, row2):
    for val in row:
      result += val*val*val
  return result/6

result1=jax.jacfwd(jax.jacfwd(f_split, (0, 1)), (0, 1))(*x0)
result2=jax.jacfwd(jax.jacfwd(f))(x0)
```

**Comment by albanD** [comment]
> This inserts an unnecessary singleton dimension in the case when only one is a tuple

I agree, this will only be added if there are multiple elements in that dimensions.

> result1[i][j][k][l]=result2[i][j][k][l] 

Hmm does it?
```python
import jax
import numpy as np


x0 = np.array([[1.,2.],[3.,4.]])

def f(x):
  result = 0
  for row in x:
    for val in row:
      result += val*val*val
  return result/6

# same as above, but each row is a separate argument
def f_split(row1, row2):
  result = 0
  for row in (row1, row2):
    for val in row:
      result += val*val*val
  return result/6

result1=jax.jacfwd(jax.jacfwd(f_split, (0, 1)), (0, 1))(*x0)
result2=jax.jacfwd(jax.jacfwd(f))(x0)

print(result1[0][1][0][1]) # prints 0.
print(result2[0][1][0][1]) # prints 2.
```
If you give 1 arg vs 2 arg, they do interleave argument number indexing with Tensor dimension indexing.

Note that the API above explicitly forbids list / dictionary. We may add it later if people really want it, but NestedTensor + NamedTensor should give most of their benefit without the need to leave the Tensor world.

**Comment by yaroslavvb** [comment]
Note that if you replace `x0=np.array([[1.,2.],[3.,4.]])` with `x0=[[1.,2.],[3.,4.]]`, then two results are the same.

```
import jax
import numpy as np


x0 = [[1.,2.],[3.,4.]]

def f(x):
  result = 0
  for row in x:
    for val in row:
      result += val*val*val
  return result/6

# same as above, but each row is a separate argument
def f_split(row1, row2):
  result = 0
  for row in (row1, row2):
    for val in row:
      result += val*val*val
  return result/6

result1=jax.jacfwd(jax.jacfwd(f_split, (0, 1)), (0, 1))(*x0)
result2=jax.jacfwd(jax.jacfwd(f))(x0)

print(result1[0][1][0][1]) # prints 2.0
print(result2[0][1][0][1]) # prints 2.0
```

I guess the question is whether to treat a tuple of 1D tensors in the same manner as a 2D (ragged) Tensor or to produce transposed results for two cases. The main case for transposition seems to be  efficiency since current autograd engine will produce transposed order.

**Comment by albanD** [comment]
Hi,

Yes, but the API proposed here won't allow a list as input. So only the case where numpy arrays are given is relevant.
I think we agree. I'll do a first draft to see how it looks.

**Comment by albanD** [comment]
Small question about the behavior:
How should we handle the case where the output of the user-function does not depend on the input (either has requires_grad=False or calling `grad(..., allow_unused=False)` raise an error) ?

The current behavior in the gradient functions are the following:
- Returns a Tensor full of 0s
- Returns None
- Raise an error

Following discussion with @zou3519 , my current idea is to do:
- Returns a Tensor full of 0s
- In anomaly mode, throw a warning when we detect this.

Any thoughts?

**Comment by yaroslavvb** [comment]
I'm leaning towards Option 1.

The reason is that concept of "not a function of inputs" depends on computation graph structure, and that's a volatile implementation detail.

Consider following two cases:

```
grad(0*x, x)
grad(0., x)
```

In case above, `0*x` and `0.`, are mathematically equivalent so PyTorch could have an optimizer pass that turns one into another. Alternatively `0*x` could come from a user-provided function, and this optimization may be done by the user.

TensorFlow which made distinction between 0 and None, and it was confusing users -- https://github.com/tensorflow/tensorflow/issues/783

**Comment by albanD** [comment]
After more thinking, I would be leaning towards doing a behavior similar to `allow_unused` in `autograd.grad()`:
- By default raise an error
- If the flag is set, returns a Tensor full of 0s.

My feeling is that for most users, the independence of the output wrt the input is going to be a bug/misconception more than an expected behavior and it is nice to tell them about it.
We still have the flag so that users that are aware of this will be able to get the result they want.

Does that sound reasonable? Or you think that most users don't want to be warned about such case and will be happy with never being aware of it?

**Comment by yaroslavvb** [comment]
The underlying issue is how to handle derivative of a constant expression. As I understand the argument above -- if user calls grad on a constant expression, they probably didn't realize it is a constant expression (or they'd hardcode zero), hence we should alert the user.

Getting derivative of a constant expression and sticking it into optim.SGD is surely a user error, but what if you take a second derivative of linear expression? I think that should be fine, but then you can't do it with grad(grad(linear_function)). It's similar to annoyance with TensorFlow special handling of constant derivatives -- you ended up with `None` blocks in your Hessian.

Also, recognizing constant expressions can be tricky, here are some examples:

```
grad(1, x)
grad(x-x, x)
grad(jit(x-x), x)
grad(sin(x)**2+cos(x)**2, x)
grad(jit(sin(x)**2+cos(x)**2), x)
```

Note that Jax treats all expressions above uniformly -- they evaluate to 0.

I guess the issue is whether you want torch.autograd to specialize to existing torch.nn users, or be consistent in a more general setting

**Comment by albanD** [comment]
After multiple discussion. The current conclusion is that the are going to return Tensors full of 0s and add a flag `strict=False` in each function that can be set to True if the user wants to ensure that all the computed values are properly connected.

**Comment by albanD** [comment]
The PR is now merged.

**Comment by slerman12** [comment]
Do gradients flow through the functional hessian operation? I know that "create graph" enables gradients to flow through at least up to the inputs passed into the function, but what if the inputs are themselves a function of some upstream network? Would gradients flow to the original source? (Let me know if I need to clarify this looser English with more formal language and an example).

**Comment by albanD** [comment]
Hi,

Yes gradients will flow back if you use `create_graph=True`.

If you have further question on this, please use the forum here to reduce noise on github: https://discuss.pytorch.org/
